[AMDGPU] Extend zero initialization of return values for TFE

dstutt opened the pull request:
buffer_load instructions that use TFE also need to zero initialize return values similar to how the image instructions currently work. Add support for this with standard zero init of all results + zero init of just TFE flag when enable-prt-strict-null subtarget feature is disabled.

dstutt committed:
[AMDGPU] Extend zero initialization of return values for TFE

buffer_load instructions that use TFE also need to zero initialize return values
similar to how the image instructions currently work.
Add support for this with standard zero init of all results + zero init of just TFE flag when
enable-prt-strict-null subtarget feature is disabled.

dstutt committed:
clang format fixes

dstutt commented:
There's no global isel support implemented yet - but I thought that could maybe be a separate change? (I'm not sure about the best approach currently).

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -15033,57 +15033,66 @@ SDNode *SITargetLowering::PostISelFolding(MachineSDNode *Node,
+      return;
 
-  const DebugLoc &DL = MI.getDebugLoc();
+    // At least one of TFE or LWE are non-zero
```
Missing globalisel handling

dstutt replied:
Yes, see the previous comment.

dstutt replied:
GlobalISel now supported

dstutt commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -15033,57 +15033,66 @@ SDNode *SITargetLowering::PostISelFolding(MachineSDNode *Node,
+    if (DstSize < InitIdx)
+      return;
+  } else if (TII->isMUBUF(MI) && AMDGPU::getMUBUFHasTFE(MI.getOpcode())) {
+    uint32_t DstSize =
```
DstSize is not required here - assign directly to InitIdx

dstutt replied:
Done

dstutt committed:
Remove superfluous DstSize

arsenm commented:
> There's no global isel support implemented yet - but I thought that could maybe be a separate change? (I'm not sure about the best approach currently).

I think it's best to do them together, otherwise they bifurcate

dstutt commented:
> > There's no global isel support implemented yet - but I thought that could maybe be a separate change? (I'm not sure about the best approach currently).
> 
> I think it's best to do them together, otherwise they bifurcate

Yes, I thought you might say that :)
I'll take another look.
Any suggestions? Image support is done in AMDGPUInstructionSelector.cpp:selectImageIntrinsic - so not appropriate for the buffer instructions - but it seems like overkill to implement a selectBufferLoadTfe when selectImpl does a good job already (but maybe that's the best approach??)

arsenm commented:
> Yes, I thought you might say that :) I'll take another look. Any suggestions? Image support is done in AMDGPUInstructionSelector.cpp:selectImageIntrinsic - so not appropriate for the buffer instructions - but it seems like overkill to implement a selectBufferLoadTfe when selectImpl does a good job already (but maybe that's the best approach??)

The handling of all of these intrinsics is split between legalizeImageIntrinsic / legalizeBuffer* and the select* functions. Anything changing the register types used should be done in the legalizer

kosarev commented on llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:
```diff
@@ -466,6 +467,11 @@ bool getMUBUFIsBufferInv(unsigned Opc) {
   return Info ? Info->IsBufferInv : false;
 }
 
+bool getMUBUFHasTFE(unsigned Opc) {
```
Should this be `getMUBUFIsTFE()`?

dstutt replied:
Maybe - TFE stands for "texture fault enable" so HasTFE feels more logical than IsTFE.

dstutt replied:
I left it as HasTFE - unless you feel strongly that it should be IsTFE?

kosarev replied:
Looks very minor, so I guess should be fine either way. The doubt was, HasTFE might be misleading in presence of `getMUBUFHasSoffset()`, etc. which have a different meaning.

dstutt replied:
I decided to go with getMUBUFTfe - since that mirrors the name of the field.

dstutt committed:
Add support for GlobalISel

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:
```diff
@@ -3321,6 +3325,29 @@ bool AMDGPUInstructionSelector::selectBufferLoadLds(MachineInstr &MI) const {
+    return false;
+
+  // Retrieve the newly created instruction
+  MachineInstr &NewMI = *(std::prev(MII));
```
Extra parens

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:
```diff
@@ -3321,6 +3325,29 @@ bool AMDGPUInstructionSelector::selectBufferLoadLds(MachineInstr &MI) const {
+  auto MII = MachineBasicBlock::iterator(MI);
+  MII = std::next(MII);
+
+  // Standard selection works fine, use that to generate the instruction
+  if (!selectImpl(MI, *CoverageInfo))
+    return false;
```
This is pretty ugly and you're making assumptions about what the generated selector does. You don't have the equivalent handling in the DAG path? I assume this is done in the post-isel hook. FinalizeISel still runs the same EmitInstrWithCustomInserter, so the same code should work for both paths without this extra post-processing?

dstutt replied:
Good idea - I've re-factored to use EmitInstrWithCustomInserter.

dstutt committed:
Use CustomerInserter approach

This means both DAG Isel and GlobalISel use the same code to zero the results for tfe instructions
Additional test update due to discovery that tfe/lwe wasn't zeroing for image_msaa with previous
implementation.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -15033,57 +15038,64 @@ SDNode *SITargetLowering::PostISelFolding(MachineSDNode *Node,
+  const DebugLoc &DL = MI.getDebugLoc();
 
   // Create a register for the initialization value.
   Register PrevDst = MRI.createVirtualRegister(TII->getOpRegClass(MI, DstIdx));
```
This could just clone the register instead of querying the regclass

dstutt replied:
Done.

dstutt committed:
Clone register rather than create from RC

jayfoad commented on llvm/lib/Target/AMDGPU/BUFInstructions.td:
```diff
@@ -119,6 +119,7 @@ class MTBUF_Real <MTBUF_Pseudo ps, string real_name = ps.Mnemonic> :
 
   let isPseudo = 0;
   let isCodeGenOnly = 0;
+  let hasPostISelHook = 1;
```
Can this be removed now? Setting it on a Real seems weird anyway.

dstutt replied:
I missed this one - it shouldn't be there. I'll remove it.

jayfoad commented on llvm/lib/Target/AMDGPU/BUFInstructions.td:
```diff
@@ -390,6 +392,7 @@ class MUBUF_Invalidate <string opName, SDPatternOperator node = null_frag> :
   let has_slc     = 0;
   let has_sccb    = 0;
   let sccb_value  = 0;
+  let tfe         = 0;
```
Could maybe set it to 0 in the base class (BUF_Pseudo) instead of in three different subclasses?

dstutt replied:
Yes. Done.

jayfoad commented on llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.struct.buffer.load.format.ll:
```diff
@@ -1,6 +1,7 @@
 ; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
 ; RUN: llc -global-isel -mtriple=amdgcn-mesa-mesa3d -mcpu=fiji -stop-after=instruction-select -verify-machineinstrs -o - %s | FileCheck -check-prefix=GFX8 %s
 ; RUN: llc -global-isel -mtriple=amdgcn-mesa-mesa3d -mcpu=gfx1200 -stop-after=instruction-select -verify-machineinstrs -o - %s | FileCheck -check-prefix=GFX12 %s
+; Note that TFE instructions don't have the result initialization to zero due to stopping before finalize-isel - which is where that's inserted
```
Maybe change it to stop-after=finalize-isel?

dstutt replied:
I tried that, but that ended up making the output quite different for the rest of the testing.
Some of the other tests verify the zero initialization for global-isel so i thought it was easier to leave these as-is and add the note.

I'm somewhat ambivalent - so I'm happy to do whatever you would prefer.

jayfoad replied:
Seems fine as-is.

jayfoad commented on llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.msaa.load.ll:
```diff
@@ -22,18 +22,36 @@ main_body:
+; GFX11-NEXT:    s_delay_alu instid0(VALU_DEP_3) | instskip(NEXT) | instid1(VALU_DEP_3) ; encoding: [0x93,0x01,0x87,0xbf]
+; GFX11-NEXT:    v_dual_mov_b32 v2, v10 :: v_dual_mov_b32 v3, v11 ; encoding: [0x0a,0x01,0x10,0xca,0x0b,0x01,0x02,0x02]
+; GFX11-NEXT:    v_mov_b32_e32 v4, v12 ; encoding: [0x0c,0x03,0x08,0x7e]
+; GFX11-NEXT:    image_msaa_load v[0:4], v[5:7], s[0:7] dmask:0x2 dim:SQ_RSRC_IMG_2D_MSAA unorm tfe lwe ; encoding: [0x98,0x02,0x60,0xf0,0x05,0x00,0x60,0x00]
```
Not your fault, but this code is terrible. It's zeroing v[8:12] and then copying that into v[0:4].

jayfoad reviewed (commented):
Looks good overall.

dstutt committed:
Minor fixups after review feedback
